File: Tree/C/M654-maximum-binary-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct TreeNode* constructMaximumBinaryTree(int* nums, int numsSize) {
    int i;
    int m = -1;
    int mi = -1;
    for(i = 0; i < numsSize; i++){
        if(nums[i] > m){
            m = nums[i];
            mi = i;
        }
    }
    
    struct TreeNode *t = (struct TreeNode *)malloc(sizeof(struct TreeNode));
    t->val = m;
    if(mi == 0)
        t->left = NULL;
    else{
        int ln[mi];
        for(i = 0; i < mi; i++)
            ln[i] = nums[i];
        t->left = constructMaximumBinaryTree(ln, i);
    }
    if(mi == numsSize - 1)
        t->right = NULL;
    else{
        int rn[numsSize - mi - 1];
        for(i = 0; i < numsSize - mi - 1; i++)
            rn[i] = nums[mi + i + 1];
        t->right = constructMaximumBinaryTree(rn, i);
    }
    
    return t;
}

File: Tree/Python3/M654-maximum-binary-tree.py
# time:  O(n)
# space: O(n)

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def constructMaximumBinaryTree(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        m = max(nums)
        i = nums.index(m)
        
        t = TreeNode(m)
        if i != 0:
            t.left = self.constructMaximumBinaryTree(nums[:i])
        if i != len(nums)-1:
            t.right = self.constructMaximumBinaryTree(nums[i+1:])
        
        return t
		
	# faster solution
	def constructMaximumBinaryTree(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        s = []
        for num in nums:
            cur = TreeNode(num)
            while s and s[-1].val < num:
                cur.left = s.pop()
            if s:
                s[-1].right = cur
            s.append(cur)
        return s[0]